File: Emulator/src/main/content/global/skill/summoning/familiar/npc/VoidFamiliarNPC.java
package content.global.skill.summoning.familiar.npc;

import content.global.skill.summoning.familiar.Familiar;
import content.global.skill.summoning.familiar.FamiliarSpecial;
import content.global.skill.summoning.familiar.Forager;
import core.cache.def.impl.NPCDefinition;
import core.game.interaction.OptionHandler;
import core.game.node.Node;
import core.game.node.entity.combat.BattleState;
import core.game.node.entity.combat.equipment.WeaponInterface;
import core.game.node.entity.player.Player;
import core.game.node.entity.skill.SkillBonus;
import core.game.node.entity.skill.Skills;
import core.game.node.item.Item;
import core.game.system.task.Pulse;
import core.game.world.GameWorld;
import core.game.world.map.Location;
import core.game.world.update.flag.context.Animation;
import core.game.world.update.flag.context.Graphics;
import core.plugin.ClassScanner;
import core.plugin.Initializable;
import core.plugin.Plugin;

/**
 * The type Void familiar npc.
 */
@Initializable
public final class VoidFamiliarNPC implements Plugin<Object> {

    private static final Item[] ITEMS = new Item[]{new Item(434), new Item(440), new Item(453), new Item(444), new Item(447)};

    @Override
    public Plugin<Object> newInstance(Object arg) throws Throwable {
        ClassScanner.definePlugin(new VoidRavagerNPC());
        ClassScanner.definePlugin(new VoidShifterNPC());
        ClassScanner.definePlugin(new VoidSpinnerNPC());
        ClassScanner.definePlugin(new VoidTorcherNPC());
        return this;
    }

    @Override
    public Object fireEvent(String identifier, Object... args) {
        return null;
    }

    /**
     * Call to arms boolean.
     *
     * @param familiar the familiar
     * @param special  the special
     * @return the boolean
     */
    public boolean callToArms(content.global.skill.summoning.familiar.Familiar familiar, content.global.skill.summoning.familiar.FamiliarSpecial special) {
        final Player owner = familiar.getOwner();
        owner.lock();
        GameWorld.getPulser().submit(new Pulse(1, owner) {
            int counter;

            @Override
            public boolean pulse() {
                switch (++counter) {
                    case 1:
                        owner.visualize(Animation.create(8136), Graphics.create(1503));
                        break;
                    case 3:
                        owner.unlock();
                        owner.getProperties().setTeleportLocation(Location.create(2659, 2658, 0));
                        owner.visualize(Animation.create(8137), Graphics.create(1502));
                        return true;
                }
                return false;
            }
        });
        return true;
    }

    /**
     * The type Void ravager npc.
     */
    public final class VoidRavagerNPC extends Forager {

        /**
         * Instantiates a new Void ravager npc.
         */
        public VoidRavagerNPC() {
            this(null, 7370);
        }

        /**
         * Instantiates a new Void ravager npc.
         *
         * @param owner the owner
         * @param id    the id
         */
        public VoidRavagerNPC(Player owner, int id) {
            super(owner, id, 2700, 12818, 3, WeaponInterface.STYLE_AGGRESSIVE, ITEMS);
            boosts.add(new SkillBonus(Skills.MINING, 1));
        }

        @Override
        public content.global.skill.summoning.familiar.Familiar construct(Player owner, int id) {
            return new VoidRavagerNPC(owner, id);
        }

        @Override
        protected boolean specialMove(content.global.skill.summoning.familiar.FamiliarSpecial special) {
            return callToArms(this, special);
        }

        @Override
        public int[] getIds() {
            return new int[]{7370, 7371};
        }

    }

    /**
     * The type Void shifter npc.
     */
    public class VoidShifterNPC extends content.global.skill.summoning.familiar.Familiar {

        /**
         * Instantiates a new Void shifter npc.
         */
        public VoidShifterNPC() {
            this(null, 7367);
        }

        /**
         * Instantiates a new Void shifter npc.
         *
         * @param owner the owner
         * @param id    the id
         */
        public VoidShifterNPC(Player owner, int id) {
            super(owner, id, 9400, 12814, 3, WeaponInterface.STYLE_ACCURATE);
        }

        @Override
        public content.global.skill.summoning.familiar.Familiar construct(Player owner, int id) {
            return new VoidShifterNPC(owner, id);
        }

        @Override
        public void adjustPlayerBattle(BattleState state) {
            super.adjustPlayerBattle(state);
            int percentage = (int) (owner.getSkills().getStaticLevel(Skills.HITPOINTS) * 0.10);
            if (owner.getSkills().getLifepoints() < percentage) {
                owner.getProperties().setTeleportLocation(Location.create(2659, 2658, 0));
            }
        }

        @Override
        protected boolean specialMove(content.global.skill.summoning.familiar.FamiliarSpecial special) {
            return callToArms(this, special);
        }

        @Override
        public int[] getIds() {
            return new int[]{7367, 7368};
        }

    }

    /**
     * The type Void spinner npc.
     */
    public class VoidSpinnerNPC extends content.global.skill.summoning.familiar.Familiar {

        private int healDelay;

        /**
         * Instantiates a new Void spinner npc.
         */
        public VoidSpinnerNPC() {
            this(null, 7333);
        }

        /**
         * Instantiates a new Void spinner npc.
         *
         * @param owner the owner
         * @param id    the id
         */
        public VoidSpinnerNPC(Player owner, int id) {
            super(owner, id, 2700, 12780, 3, WeaponInterface.STYLE_DEFENSIVE);
        }

        @Override
        public void handleFamiliarTick() {
            super.handleFamiliarTick();
            if (healDelay < GameWorld.getTicks()) {
                getSkills().heal(1);
                healDelay = GameWorld.getTicks() + 25;
            }
        }

        @Override
        public content.global.skill.summoning.familiar.Familiar construct(Player owner, int id) {
            return new VoidSpinnerNPC(owner, id);
        }

        @Override
        protected boolean specialMove(content.global.skill.summoning.familiar.FamiliarSpecial special) {
            return callToArms(this, special);
        }

        @Override
        public int[] getIds() {
            return new int[]{7333, 7334};
        }

    }

    /**
     * The type Void torcher npc.
     */
    public class VoidTorcherNPC extends content.global.skill.summoning.familiar.Familiar {

        /**
         * Instantiates a new Void torcher npc.
         */
        public VoidTorcherNPC() {
            this(null, 7351);
        }

        /**
         * Instantiates a new Void torcher npc.
         *
         * @param owner the owner
         * @param id    the id
         */
        public VoidTorcherNPC(Player owner, int id) {
            super(owner, id, 9400, 12798, 3, WeaponInterface.STYLE_CAST);
        }

        @Override
        public content.global.skill.summoning.familiar.Familiar construct(Player owner, int id) {
            return new VoidTorcherNPC(owner, id);
        }

        @Override
        public void configureFamiliar() {
            ClassScanner.definePlugin(new OptionHandler() {

                @Override
                public Plugin<Object> newInstance(Object arg) throws Throwable {
                    for (int i : getIds()) {
                        NPCDefinition.forId(i).getHandlers().put("option:strike", this);
                    }
                    return this;
                }

                @Override
                public boolean handle(Player player, Node node, String option) {
                    final content.global.skill.summoning.familiar.Familiar familiar = (Familiar) node;
                    if (!player.getFamiliarManager().isOwner(familiar)) {
                        return true;
                    }
                    // TODO:
                    return true;
                }

            });
        }

        @Override
        protected boolean specialMove(FamiliarSpecial special) {
            return callToArms(this, special);
        }

        @Override
        public int[] getIds() {
            return new int[]{7351, 7352};
        }

    }
}

File: Emulator/src/main/content/global/skill/summoning/familiar/npc/VoidFamiliarNPC.kt
package content.global.skill.summoning.familiar.npc

import content.global.skill.summoning.familiar.Familiar
import content.global.skill.summoning.familiar.FamiliarSpecial
import content.global.skill.summoning.familiar.Forager
import core.cache.def.impl.NPCDefinition
import core.game.interaction.OptionHandler
import core.game.node.Node
import core.game.node.entity.combat.BattleState
import core.game.node.entity.combat.equipment.WeaponInterface
import core.game.node.entity.player.Player
import core.game.node.entity.skill.SkillBonus
import core.game.node.entity.skill.Skills
import core.game.node.item.Item
import core.game.system.task.Pulse
import core.game.world.GameWorld
import core.game.world.GameWorld.Pulser
import core.game.world.map.Location
import core.game.world.update.flag.context.Animation
import core.game.world.update.flag.context.Graphics
import core.plugin.ClassScanner.definePlugin
import core.plugin.Initializable
import core.plugin.Plugin
import shared.consts.Items
import shared.consts.NPCs

@Initializable
class VoidFamiliarNPC : Plugin<Any?> {
    @Throws(Throwable::class)
    override fun newInstance(arg: Any?): Plugin<Any?> {
        definePlugin(VoidRavagerNPC())
        definePlugin(VoidShifterNPC())
        definePlugin(VoidSpinnerNPC())
        definePlugin(VoidTorcherNPC())
        return this
    }

    override fun fireEvent(identifier: String, vararg args: Any): Any? {
        return null
    }

    fun callToArms(familiar: Familiar, special: FamiliarSpecial?): Boolean {
        val owner = familiar.owner
        owner.lock()
        Pulser.submit(object : Pulse(1, owner) {
            var counter: Int = 0

            override fun pulse(): Boolean {
                when (++counter) {
                    1 -> owner.visualize(Animation.create(8136), Graphics.create(1503))
                    3 -> {
                        owner.unlock()
                        owner.properties.teleportLocation = Location.create(2659, 2658, 0)
                        owner.visualize(Animation.create(8137), Graphics.create(1502))
                        return true
                    }
                }
                return false
            }
        })
        return true
    }


    inner class VoidRavagerNPC @JvmOverloads constructor(owner: Player? = null, id: Int = NPCs.VOID_RAVAGER_7370) :
        Forager(owner, id, 2700, Items.VOID_RAVAGER_POUCH_12818, 3, WeaponInterface.STYLE_AGGRESSIVE, *ITEMS) {


        init {
            boosts.add(SkillBonus(Skills.MINING, 1.0))
        }

        override fun construct(owner: Player, id: Int): Familiar {
            return VoidRavagerNPC(owner, id)
        }

        override fun specialMove(special: FamiliarSpecial): Boolean {
            return callToArms(this, special)
        }

        override fun getIds(): IntArray {
            return intArrayOf(NPCs.VOID_RAVAGER_7370, NPCs.VOID_RAVAGER_7371)
        }
    }


    inner class VoidShifterNPC
    @JvmOverloads constructor(owner: Player? = null, id: Int = NPCs.VOID_SHIFTER_7367) :
        Familiar(owner, id, 9400, Items.VOID_SHIFTER_POUCH_12814, 3, WeaponInterface.STYLE_ACCURATE) {
        override fun construct(owner: Player, id: Int): Familiar {
            return VoidShifterNPC(owner, id)
        }

        override fun adjustPlayerBattle(state: BattleState) {
            super.adjustPlayerBattle(state)
            val percentage = (owner.getSkills().getStaticLevel(Skills.HITPOINTS) * 0.10).toInt()
            if (owner.getSkills().lifepoints < percentage) {
                owner.properties.teleportLocation = Location.create(2659, 2658, 0)
            }
        }

        override fun specialMove(special: FamiliarSpecial): Boolean {
            return callToArms(this, special)
        }

        override fun getIds(): IntArray {
            return intArrayOf(NPCs.VOID_SHIFTER_7367, NPCs.VOID_SHIFTER_7368)
        }
    }


    inner class VoidSpinnerNPC
    @JvmOverloads constructor(owner: Player? = null, id: Int = NPCs.VOID_SPINNER_7333) :
        Familiar(owner, id, 2700, Items.VOID_SPINNER_POUCH_12780, 3, WeaponInterface.STYLE_DEFENSIVE) {
        private var healDelay = 0

        public override fun handleFamiliarTick() {
            super.handleFamiliarTick()
            if (healDelay < GameWorld.ticks) {
                getSkills().heal(1)
                healDelay = GameWorld.ticks + 25
            }
        }

        override fun construct(owner: Player, id: Int): Familiar {
            return VoidSpinnerNPC(owner, id)
        }

        override fun specialMove(special: FamiliarSpecial): Boolean {
            return callToArms(this, special)
        }

        override fun getIds(): IntArray {
            return intArrayOf(NPCs.VOID_SPINNER_7333, NPCs.VOID_SPINNER_7334)
        }
    }


    inner class VoidTorcherNPC
    @JvmOverloads constructor(owner: Player? = null, id: Int = NPCs.VOID_TORCHER_7351) :
        Familiar(owner, id, 9400, Items.VOID_TORCHER_POUCH_12798, 3, WeaponInterface.STYLE_CAST) {
        override fun construct(owner: Player, id: Int): Familiar {
            return VoidTorcherNPC(owner, id)
        }

        public override fun configureFamiliar() {
            definePlugin(object : OptionHandler() {
                @Throws(Throwable::class)
                override fun newInstance(arg: Any?): Plugin<Any> {
                    for (i in ids) {
                        NPCDefinition.forId(i).handlers["option:strike"] = this
                    }
                    return this
                }

                override fun handle(player: Player, node: Node, option: String): Boolean {
                    val familiar = node as Familiar
                    if (!player.familiarManager.isOwner(familiar)) {
                        return true
                    }
                    return true
                }
            })
        }

        override fun specialMove(special: FamiliarSpecial): Boolean {
            return callToArms(this, special)
        }

        override fun getIds(): IntArray {
            return intArrayOf(NPCs.VOID_TORCHER_7351, NPCs.VOID_TORCHER_7352)
        }
    }

    companion object {
        private val ITEMS = arrayOf(
            Item(Items.CLAY_434),
            Item(Items.IRON_ORE_440),
            Item(Items.COAL_453),
            Item(Items.GOLD_ORE_444),
            Item(Items.MITHRIL_ORE_447)
        )
    }
}
